rfcs/0157-more-granular-control.md

```diff
@@ -0,0 +1,241 @@
+
+With the introduction of `nix run` and `nix build`, all current implementations should transition to using these first-party nix commands instead of wrappers like nixos-rebuild, nixos-generate, and home-manager.
+
+Operating everything under flake also opens up various opportunities, such as building the latest nixos iso image with or without plasma/gnome directly from the nixpkgs repository if its derivation is exposed in flake.nix (for example, `nix build nixpkgs#isoInstallerImageWithGnome`).
+
+### CLI improvements
+
+As explained earlier in the motivation section, the binary cache has no direct access, and using `nix-store --import {exported}` does not automatically pull missing dependencies from the binary cache in case the export does not include the whole closure.
+
+Allowing direct binary-cache access, such as querying information and pulling specific paths into the local `/nix/store`, would be a significant enhancement.
+
+### (optional) stdenv improvements
+
+Optionally, we could accompany this with a simpler, low-abstraction alternative to stdenv, or an opt-in C compiler setup.
+
+It is worth noting that `stdenv.mkDerivation` pulls gcc and autotools, but not all builds require these dependencies. Some simple shell scripts and builds in other languages redundantly pull these dependencies as part of the build, even when builtins.derivation provided by nix language could be sufficient.
+
+The automatic header include and ldflag magic provided by stdenv often leads to a "just-works" mentality, causing us to become overly dependent on it.
+
+
+# Examples and Interactions
+[examples-and-interactions]: #examples-and-interactions
+
+### Everything is opt-in!
+
+Decoupling the system from fixed parts can be done by availing configuration options under the NixOS modules:
+
+```nix
+{
+  # bootloader.grub.enable = false;
+  bootloader.gummiboot.enable = true;
+  init.runit.enable = true;
+  boot.kernelPackages = pkgs.linuxPackages_latest;
+  boot.baseFiles = pkgs.baseFiles_latest;
+  # Other system configurations can be added here  ...
+}
+```
+
+This is an example constituting how one can easily replace grub with gummiboot.
+
+By providing this level of granularity, users can precisely control their system components, opting for their preferred choices while excluding unnecessary defaults. This empowers users to create tailored configurations and enhances the flexibility of their NixOS setups.
+
+### Emphasis on flake over shell script wrappers
+
+[nixos-rebuild](https://github.com/NixOS/nixpkgs/blob/master/pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.sh) is nothing but a complex shell script, that has to be present in the system beforehand, which basically calls `nix-build` under the hood, building a derivation, and then running the built derivation. All the options can be instead defined in nix language, in the flake definition itself, and the output derivation can be built using the nix command directly.
```

Comment:
This would effectively lock out everyone who doesn't want to use experimental features. Also, the scripts have an existing shell completion whereas there's no proper completion system out there that completes nix code for me.

Reply:
The configuration options (including personal comments) are more flexible than remembering / checking history.

Surely there's no alternative than shell completion at the moment, but you can look up for `nix repl`, all the auto-completion work there, you can import your flake `:lf .` and then you can query the nixos modules using tab completion. As documentation improves, flake usage improves as well.

And as said this will be more flexible too in the future because flake outputs are accessible easily, remotely:

> Operating everything under flake also opens up various opportunities, such as building the latest nixos iso image with or without plasma/gnome directly from the nixpkgs repository if its derivation is exposed in flake.nix (for example, `nix build nixpkgs#isoInstallerImageWithGnome`).

And its not to remove the script wrappers altogether, but just like nixos-generators add an alternative way to configure with flakes which could in future stablize and script wrappers be deprecated.

Reply:
> The configuration options (including personal comments) are more flexible than remembering / checking history.

Can you give an example perhaps about which options you'd like to see in Nix rather than as CLI options? I wouldn't want to e.g. change the flake whenever I exchange `switch`/`test`/etc whenever needed for instance?

Reply:
I'm not suggesting to put these in flakes

```
      switch|boot|test|build|edit|dry-build|dry-run|dry-activate|build-vm|build-vm-with-bootloader)
```
these are the options that are freqently changed, and can be added as different outputs (`.dry`|`.test`) suffixes.

but rather options such as

```
      --install-grub)
      --install-bootloader)
      --no-build-nix)
      --use-substitutes|--substitute-on-destination|-s)
      -I|--max-jobs|-j|--cores|--builders|--log-format)
      -j*|--quiet|--print-build-logs|-L|--no-build-output|-Q| --show-trace|--keep-going|-k|--keep-failed|-K|--fallback|--refresh|--repair|--impure|--offline|--no-net)
      --option)
      --fast)
      --profile-name|-p)
# and many more...
```
are generally fixed, always, in every switch/build, these can be specified as a nix configuration option instead in the flake output.

Reply:
Uhm,  on my system I sometimes change the state of  `-j` and `-k` depending on wokflow context of preparing the update. It's not mainline NixOS but I would expect _some_ users to do the same on NixOS.

Reply:
> This would effectively lock out everyone who doesn't want to use experimental features. Also, the scripts have an existing shell completion whereas there's no proper completion system out there that completes nix code for me.

I think it's time to not feature flag nix-command and flakes anymore. Everyone will use these features anyway, this extra step only annoys everyone.

Reply:
Flakes are still not stable enough.